File: FinalProject/1_IntegerPartition/Baitoan1/Ferrers.py
def gen_ordered_partitions(cur, k, n_remain, part, res):
    if cur == k:
        if n_remain == 0:
            res.append(part.copy())
        return
    min_val = 1
    max_val = n_remain - (k - cur - 1)
    for val in range(min_val, max_val + 1):
        part[cur] = val
        gen_ordered_partitions(cur + 1, k, n_remain - val, part, res)

def gen_partitions(cur, k, n_remain, part, res):
    if cur == k:
        if n_remain == 0:
            res.append(part.copy())
        return
    start_val = n_remain if cur == 0 else min(n_remain, part[cur - 1])
    for val in range(start_val, 0, -1):
        part[cur] = val
        gen_partitions(cur + 1, k, n_remain - val, part, res)

def ferrers(part):
    for val in part:
        print('*' * val)

def ferrers_transpose(part):
    max_height = max(part)
    k = len(part)
    for i in range(1, max_height + 1):
        for j in range(k):
            print('*' if part[j] >= i else ' ', end='')
        print()

n, k = map(int, input().split())
if k > n or k <= 0 or n <= 0:
    print("Khong co phan khoach")
else:
    partitions = []
    part = [0] * k

    gen_partitions(0, k, n, part, partitions)
    # gen_ordered_partitions(0, k, n, part, partitions)

    print(f"Co p_{k}({n}) = {len(partitions)} phan khoach cua {n} co {k} phan tu:\n")

    cnt = 1
    for p in partitions:
        print(f"Phan khoach {cnt}: ({', '.join(map(str, p))})\n")
        print("Bieu do Ferrers:")
        ferrers(p)
        print()
        print("Bieu do Ferrers chuyen vi:")
        ferrers_transpose(p)
        print("\n---------------------\n")
        cnt += 1

File: FinalProject/1_IntegerPartition/Baitoan1/Ferrers.cpp
#include <bits/stdc++.h>
using namespace std;

//hàm đệ quy sinh phân hoạch k phần tử >=1, tổng n
//cur: index hiện tại
//nRemain: tổng còn lại cần phân phối
//part: phân hoạch hiện tại
//res: tất cả phân hoạch

//có xét thứ tự
void genOrderedPartitions(int cur, int k, int nRemain, vector<int>& part, vector<vector<int>>& res) {
    if (cur==k) {
        if (nRemain==0) res.push_back(part);
        return;
    }
    //vì mỗi phần tử >= 1, nên phần tử tại cur có thể nhận giá trị từ 1 đến (nRemain - số phần tử còn lại)
    int minVal=1;
    int maxVal=nRemain - (k-cur-1);
    for (int val=minVal; val<=maxVal; ++val) {
        part[cur]=val;
        genOrderedPartitions(cur+1, k, nRemain-val, part, res);
    }
}

//không xét thứ tự
void genPartitions(int cur, int k, int nRemain, vector<int>& part, vector<vector<int>>& res) {
    if (cur==k) {
        if (nRemain==0) res.push_back(part);
        return;
    }
    int startVal=(cur==0) ? nRemain : min(nRemain, part[cur-1]);
    int endVal=1;
    for (int val=startVal; val>=endVal; --val) {
        part[cur] = val;
        genPartitions(cur + 1, k, nRemain-val, part, res);
    }
}

void ferrers(const vector<int>& part) {
    for (int val:part) {
        for (int i=0; i<val; ++i) cout<<'*';
        cout<<'\n';
    }
}

void ferrersTranspose(const vector<int>& part) {
    int maxHeight=*max_element(part.begin(), part.end());
    int k=part.size();

    for (int i=1; i<=maxHeight; ++i) {
        for (int j=0; j<k; ++j) {
            if (part[j]>=i) cout<<'*';
            else cout<<' ';
        }
        cout<<'\n';
    }
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n,k;
    cin>>n>>k;
    if (k>n || k<=0 || n<=0) {
        cout<<"Khong co phan khoach";
        return 0;
    }

    vector<vector<int>>partitions;
    vector<int>part(k);

    genPartitions(0, k, n, part, partitions);
    //genOrderedPartitions(0, k, n, part, partitions);

    cout<<"Co p_"<<k<<"("<<n<<") = "<<partitions.size()<<" phan khoach cua "<<n<<" co "<<k<<" phan tu:\n\n";

    int cnt=1;
    for (const auto& p : partitions) {
        cout<<"Phan khoach "<<cnt++<<": (";
        for (size_t i=0; i<p.size(); ++i) {
            cout<<p[i];
            if (i!=p.size()-1) cout<<", ";
        }
        cout<<")\n\n";

        cout<<"Bieu do Ferrers:\n";
        ferrers(p);
        cout<<'\n';

        cout<<"Bieu do Ferrers chuyen vi:\n";
        ferrersTranspose(p);
        cout << "\n---------------------\n\n";
    }
    return 0;
}
